lldb/tools/lldb-server/lldb-platform.cpp

```diff
@@ -393,125 +362,222 @@ int main_platform(int argc, char *argv[]) {
-                "no available gdbserver port for connection - dropping...\n");
-      } else {
-        const Socket *conn_socket =
-            static_cast<const Socket *>(conn->GetReadObject().get());
-        error =
-            spawn_process(progname, conn_socket, *available_port, port_offset,
-                          inferior_arguments, log_file, log_channels);
-        if (error.Fail()) {
-          {
-
-            std::lock_guard<std::mutex> guard(gdbserver_portmap_mutex);
-            gdbserver_portmap.FreePort(*available_port);
-          }
-          LLDB_LOGF(GetLog(LLDBLog::Platform), "spawn_process failed: %s",
-                    error.AsCString());
-          WithColor::error()
-              << "spawn_process failed: " << error.AsCString() << "\n";
-        }
-      }
-      // Parent doesn't need a connection to the lldb client
-      delete conn;
-
-      // Parent will continue to listen for new connections.
-      continue;
-    } else {
-      // If not running as a server, this process will not accept
-      // connections while a connection is active.
-      acceptor_up.reset();
-
-      // When not running in server mode, use all available ports
-      platform.SetPortMap(std::move(gdbserver_portmap));
+    gdb_inferior_arguments.AppendArguments(inferior_arguments);
+  }
+
+  GDBRemoteCommunicationServerPlatform platform(protocol, gdbserver_port);
+  {
+    llvm::Expected<std::vector<MainLoopBase::ReadHandleUP>> platform_handles =
+        sock_platform->Accept(
+            *g_main_loop,
+            [progname, gdbserver_port, inferior_arguments, log_file,
+             log_channels](std::unique_ptr<Socket> sock_up) {
+              // If not running as a server, this process will not accept
+              // connections while a connection is active.
+              if (g_server < 0)
+                return;
+              printf("Connection established.\n");
+              Status error =
+                  spawn_process(progname, sock_up.get(), gdbserver_port,
+                                inferior_arguments, log_file, log_channels);
+              if (error.Fail()) {
+                Log *log = GetLog(LLDBLog::Platform);
+                LLDB_LOGF(log, "spawn_process failed: %s", error.AsCString());
+                WithColor::error()
+                    << "spawn_process failed: " << error.AsCString() << "\n";
+                if (!g_server)
+                  g_main_loop->RequestTermination();
+              }
+              if (!g_server)
+                g_server = -1;
+            });
+    if (!platform_handles) {
+      printf("Failed to accept platform: %s\n",
+             llvm::toString(platform_handles.takeError()).c_str());
+      return socket_error;
     }
 
-    platform.SetConnection(std::unique_ptr<Connection>(conn));
-    client_handle(platform, inferior_arguments);
-  } while (g_server);
+    std::vector<MainLoopBase::ReadHandleUP> gdb_handles;
+    if (protocol == Socket::ProtocolTcp) {
```

Comment:
Put this into a helper function (`AcceptGdbConnectionsIfNeeded`?). Also move the listening code there.

Reply:
It is hard to combine listening and accepting code in the same helper. gdbserver_port may be updated after Listen() and it is used in the platform callback. I have created 2 helpers ListenGdbConnectionsIfNeeded() and AcceptGdbConnectionsIfNeeded().